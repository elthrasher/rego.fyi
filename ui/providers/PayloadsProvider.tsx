import React, { createContext, ReactNode, useContext, useState } from 'react';

const baseUrl = 'https://api.rego.fyi';

const dataDefault = JSON.stringify({
  requests: [{ methods: ['GET'], resources: ['/orders'] }],
  permissions: ['start_order', 'view_invoice'],
  subscriptions: ['newsletter'],
});

const regoDefault = `package entitlements

import data.requests
import data.permissions
import data.subscriptions

default allow = false

allow {
    check_entitlements[input]
}

check_entitlements[input] {
    r = requests[_]
    some i; match_with_wildcard(data.permissions, input.permissions[i])
    some j; match_with_wildcard(data.subscriptions, input.subscriptions[j])
    match_with_wildcard(r.methods, input.method)
    match_with_wildcard(r.resources, input.resource)
}

match_with_wildcard(allowed, value) {
    allowed[_] = "*"
}
match_with_wildcard(allowed, value) {
    allowed[_] = value
}`;

const tokenDefault = JSON.stringify({
  permissions: ['start_order'],
  resource: '/orders',
  method: 'GET',
  subscriptions: ['newsletter'],
});

export interface Payloads {
  data: string;
  rego: string;
  token: string;

  method: string;
  resource: string;

  status?: 'ALLOWED' | 'DENIED';
  responsePayload?: string;

  saveData: (data: string) => string;
  saveRego: (data: string) => string;
  saveToken: (data: string) => string;
  sendRequest: () => void;
  updateRequest: (method: string, resource: string) => void;
}

export const PayloadsContext = createContext<Payloads | undefined>(undefined);

export const PayloadsProvider = ({ children }: { children: ReactNode }): JSX.Element => {
  const [data, setData] = useState<string>(dataDefault);
  const [rego, setRego] = useState<string>(regoDefault);
  const [token, setToken] = useState<string>(tokenDefault);

  const [method, setMethod] = useState<string>('GET');
  const [resource, setResource] = useState<string>('/orders');

  const [status, setStatus] = useState<'ALLOWED' | 'DENIED' | undefined>();
  const [responsePayload, setResponsePayload] = useState<string | undefined>();

  const saveData = (data: string): string => {
    setData(data);
    try {
      JSON.parse(data);
      return 'valid';
    } catch (e) {
      return 'Invalid JSON';
    }
  };

  const saveRego = (rego: string): string => {
    setRego(rego);
    return 'valid';
  };

  const saveToken = (data: string): string => {
    setToken(data);
    try {
      JSON.parse(data);
      return 'valid';
    } catch (e) {
      return 'Invalid JSON';
    }
  };

  const updateRequest = (method: string, resource: string) => {
    setMethod(method);
    setResource(resource);
  };

  const sendRequest = async () => {
    const pl = btoa(JSON.stringify({ data, rego, token }));
    try {
      const response = await fetch(`${baseUrl}${resource}`, { headers: { Authorization: `Bearer ${pl}` }, method });
      if (response.status === 200) {
        setStatus('ALLOWED');
        const body = await response.json();
        setResponsePayload(JSON.stringify(body));
      } else {
        setStatus('DENIED');
        setResponsePayload(undefined);
      }
    } catch {
      setStatus('DENIED');
      setResponsePayload(undefined);
    }
  };

  return (
    <PayloadsContext.Provider
      value={{
        data,
        rego,
        token,
        method,
        resource,
        saveData,
        saveRego,
        saveToken,
        sendRequest,
        updateRequest,
        status,
        responsePayload,
      }}
    >
      {children}
    </PayloadsContext.Provider>
  );
};

export const usePayloadsContext = (): Payloads => {
  const payloadsContext = useContext(PayloadsContext);
  if (!payloadsContext) {
    throw new Error('usePayloadsContext must be used within the PayloadsContext.Provider');
  }
  return payloadsContext;
};
