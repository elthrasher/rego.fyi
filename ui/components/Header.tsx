import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import Container from '@material-ui/core/Container';
import Grid from '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';
import React from 'react';

import { usePayloadsContext } from '../providers/PayloadsProvider';

export const Header = (): JSX.Element => {
  const { status, responsePayload } = usePayloadsContext();

  return (
    <Container>
      <Grid container spacing={3}>
        <Grid item xs={3}>
          <Typography>
            <h1>rego.fyi</h1>
          </Typography>
        </Grid>
        <Grid item xs={9}>
          {status && (
            <Card>
              <CardContent>
                <Typography color="textPrimary" gutterBottom>
                  {status}
                </Typography>
                <Typography color="textSecondary">{responsePayload}</Typography>
              </CardContent>
            </Card>
          )}
        </Grid>
      </Grid>
    </Container>
  );
};
