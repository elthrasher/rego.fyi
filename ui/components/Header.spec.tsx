import React from 'react';
import renderer from 'react-test-renderer';

import { Payloads, PayloadsContext, PayloadsProvider } from '../providers/PayloadsProvider';
import { Header } from './Header';

it('renders default', () => {
  const component = renderer.create(
    <PayloadsProvider>
      <Header />
    </PayloadsProvider>,
    {
      createNodeMock: (node) => {
        return document.createElement(node.type.toString());
      },
    },
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders an ALLOWED message', () => {
  const component = renderer.create(
    <PayloadsContext.Provider value={{ status: 'ALLOWED', responsePayload: 'response' } as Payloads}>
      <Header />
    </PayloadsContext.Provider>,
    {
      createNodeMock: (node) => {
        return document.createElement(node.type.toString());
      },
    },
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders a DENIED message', () => {
  const component = renderer.create(
    <PayloadsContext.Provider value={{ status: 'ALLOWED' } as Payloads}>
      <Header />
    </PayloadsContext.Provider>,
    {
      createNodeMock: (node) => {
        return document.createElement(node.type.toString());
      },
    },
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
