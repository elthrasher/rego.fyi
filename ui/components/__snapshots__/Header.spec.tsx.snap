// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a DENIED message 1`] = `
<div
  className="MuiContainer-root MuiContainer-maxWidthLg"
>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
    >
      <p
        className="MuiTypography-root MuiTypography-body1"
      >
        <h1>
          rego.fyi
        </h1>
      </p>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-9"
    >
      <div
        className="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          className="MuiCardContent-root"
        >
          <p
            className="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextPrimary MuiTypography-gutterBottom"
          >
            ALLOWED
          </p>
          <p
            className="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders an ALLOWED message 1`] = `
<div
  className="MuiContainer-root MuiContainer-maxWidthLg"
>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
    >
      <p
        className="MuiTypography-root MuiTypography-body1"
      >
        <h1>
          rego.fyi
        </h1>
      </p>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-9"
    >
      <div
        className="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          className="MuiCardContent-root"
        >
          <p
            className="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextPrimary MuiTypography-gutterBottom"
          >
            ALLOWED
          </p>
          <p
            className="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
          >
            response
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders default 1`] = `
<div
  className="MuiContainer-root MuiContainer-maxWidthLg"
>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
    >
      <p
        className="MuiTypography-root MuiTypography-body1"
      >
        <h1>
          rego.fyi
        </h1>
      </p>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-9"
    />
  </div>
</div>
`;
